internal/logbox/logbox.go

```diff
@@ -0,0 +1,143 @@
+package logbox
+
+import (
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/charmbracelet/bubbles/spinner"
+	tea "github.com/charmbracelet/bubbletea"
+	"github.com/charmbracelet/lipgloss"
+	"github.com/fatih/color"
+	"github.com/grafana/loki/pkg/logcli/output"
+	"github.com/grafana/loki/pkg/loghttp"
+)
+
+var appStyle = lipgloss.NewStyle().Margin(0, 2, 1, 1)
+
+type Msg struct {
+	Line string
+	Done bool
+}
+
+func (r Msg) String() string {
+	return r.Line
+}
+
+// LogBox renders a scrolling box in a terminal for outputting logs without
+// filling up the whole scrollback. On the first line it shows the waitMessage
+// with a spinner in front.
+type LogBox struct {
+	height      int
+	waitMessage string
+	results     []Msg
+	quitting    bool
+	spinner     spinner.Model
+	interrupt   chan bool
+}
+
+// New initializes a LogBox. The interrupt channel will be written to on
+// ctrl+c/ctrl+d since it intercepts these.
+func New(height int, waitMessage string, interrupt chan bool) LogBox {
+	s := spinner.New(spinner.WithSpinner(spinner.MiniDot), spinner.WithStyle(lipgloss.NewStyle().Foreground(lipgloss.AdaptiveColor{})))
+	return LogBox{
+		height:      height,
+		waitMessage: waitMessage,
+		spinner:     s,
+		results:     []Msg{},
+		interrupt:   interrupt,
+	}
+}
+
+func (lb LogBox) Init() tea.Cmd {
+	return lb.spinner.Tick
+}
+
+func (lb LogBox) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+	switch msg := msg.(type) {
+	case tea.KeyMsg:
+		switch msg.String() {
+		case "ctrl+c", "ctrl+d":
+			lb.interrupt <- true
+			return lb, tea.Quit
+		default:
+			return lb, nil
+		}
+	case Msg:
+		if msg.Done {
+			lb.quitting = true
+			return lb, tea.Quit
+		}
+		if len(lb.results) == lb.height {
+			lb.results = append(lb.results[1:], msg)
+		} else {
+			lb.results = append(lb.results, msg)
+		}
+		return lb, nil
+	case spinner.TickMsg:
+		var cmd tea.Cmd
+		lb.spinner, cmd = lb.spinner.Update(msg)
+		return lb, cmd
+	default:
+		return lb, nil
+	}
+}
+
+func (lb LogBox) View() string {
+	var s string
+
```

Comment:
Do we need all these empty lines in this function?

Reply:
I think it makes things more readable :) but I removed the one after the var declaration